client/invoke.go

```diff
@@ -0,0 +1,105 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"os"
+
+	"github.com/fnproject/fn_go/provider"
+	"github.com/go-openapi/runtime/logger"
+)
+
+func Invoke(provider provider.Provider, invokeUrl string, content io.Reader, output io.Writer, method string, env []string, contentType string, includeCallID bool) error {
+
+	method = "POST"
+
+	// Read the request body (up to the maximum size), as this is used in the
+	// authentication signature
+	var req *http.Request
+	if content != nil {
+		b, err := ioutil.ReadAll(io.LimitReader(content, MaximumRequestBodySize))
+		buffer := bytes.NewBuffer(b)
+		req, err = http.NewRequest(method, invokeUrl, buffer)
+		if err != nil {
+			return fmt.Errorf("Error creating request to service: %s", err)
+		}
+	} else {
+		var err error
+		req, err = http.NewRequest(method, invokeUrl, nil)
+		if err != nil {
+			return fmt.Errorf("Error creating request to service: %s", err)
+		}
+	}
+
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	} else {
+		req.Header.Set("Content-Type", "text/plain")
+	}
+
+	if len(env) > 0 {
+		EnvAsHeader(req, env)
+	}
+
+	transport := provider.WrapCallTransport(http.DefaultTransport)
+	httpClient := http.Client{Transport: transport}
+
+	if logger.DebugEnabled() {
+		b, err := httputil.DumpRequestOut(req, content != nil)
+		if err != nil {
+			return err
+		}
+		fmt.Printf(string(b) + "\n")
+	}
+
+	resp, err := httpClient.Do(req)
+
+	if err != nil {
+		return fmt.Errorf("Error invoking fn: %s", err)
+	}
+
+	if logger.DebugEnabled() {
+		b, err := httputil.DumpResponse(resp, true)
+		if err != nil {
+			return err
+		}
+		fmt.Printf(string(b) + "\n")
+	}
+
+	// for sync calls
+	if call_id, found := resp.Header[FN_CALL_ID]; found {
```

Comment:
`callID` is golint approved.
constants also should be of camel case form and not all caps, snake cased, i.e. `FnCallID` (this convention is weird but it's the golint way)

Reply:
This is call_id due to a name clash with the callID struct used later. I'm a bit unwilling to change this stuff right now as it's used by the call route code. When we delete that, we can refactor this up nicely.